gitgud/operations.py

```diff
@@ -324,6 +326,51 @@ def get_current_tree(self):
         }
         return tree
 
+    def initialize_progress_file(self, all_skills):
+        with open(self.progress_path, 'w') as progress_file:
+            progress_data = {}
+            for skill in all_skills:
+                progress_data.update(
+                    {skill.name: {level.name: 'unvisited' for level in skill}}
+                )
+            json.dump(progress_data, progress_file)
+
+    def read_progress_file(self):
+        with open(self.progress_path) as progress_file:
+            return json.load(progress_file)
+
+    def update_progress_file(self, data):
+        progress_data = self.read_progress_file()
+        with open(self.progress_path, 'w') as progress_file:
+            progress_data.update(data)
+            json.dump(progress_data, progress_file)
+
+    def get_level_progress(self, level):
+        progress_data = self.read_progress_file()
+        return progress_data[level.skill.name][level.name]
+
+    def mark_level_complete(self, level):
+        progress_data = self.read_progress_file()
+        if (progress_data[level.skill.name][level.name] != "complete"):
+            progress_data[level.skill.name].update(
+                {level.name: "complete"}
+            )
+            self.update_progress_file(progress_data)
+
+    def mark_level_visited(self, level):
+        progress_data = self.read_progress_file()
+        progress_data[level.skill.name].update(
+            {level.name: "visited"}
+        )
+        self.update_progress_file(progress_data)
+
+    def mark_level_partial(self, level):
```

Comment:
All three of these methods should be implemented the same (mark_level_complete/visited/partial), and use the same underlying implementation. `mark_level(self, level, status)`

Reply:
I was thinking about doing it that way, but if we ever decided to (for one reason) change the structure of the JSON file, keeping thee functions separate would make it transparent to whoever was calling the function (e.g. changing the status from `"unvisited"` in the data to `"Unvisited"`).

Now that you've given the OK, though, I'll change this.